contrib/photils.lua

```diff
@@ -0,0 +1,414 @@
+        hard_min = 0,
+        soft_max = 100,
+        soft_min = 0,
+        label = _("Min Confidence Value")
+    },
+    warning = dt.new_widget("label")
+}
+
+function PHOTILS.image_changed()
+    local current_image = tostring(dt.gui.selection()[1])
+    if current_image ~= PHOTILS.tagged_image then
+        if PHOTILS.tagged_image ~= "" then
+            PHOTILS.tagged_image_has_changed()
+        end
+
+        PHOTILS.tagged_image = tostring(current_image)
+    end
+end
+
+function PHOTILS.tagged_image_has_changed()
+    GUI.warning.label = _(
+                            "The suggested tags were not generated\n for the currently selected image!")
+end
+
+function PHOTILS.paginate()
+    PHOTILS.in_pagination = true
+    local num_pages = math.ceil(#PHOTILS.tags / PHOTILS.per_page)
+    GUI.page_label.label = string.format(_("  Page %s of %s  "), PHOTILS.page,
+                                         num_pages)
+
+    if PHOTILS.page <= 1 then
+        PHOTILS.page = 1
+        GUI.prev_button.sensitive = false
+    else
+        GUI.prev_button.sensitive = true
+    end
+
+    if PHOTILS.page > num_pages - 1 then
+        PHOTILS.page = num_pages
+        GUI.next_button.sensitive = false
+    else
+        GUI.next_button.sensitive = true
+    end
+
+    local offset = ((PHOTILS.page - 1) * PHOTILS.per_page) + 1
+    local tag_index = 1
+    for i = offset, offset + PHOTILS.per_page - 1, 1 do
+        local tag = PHOTILS.tags[i]
+        GUI.tag_box[tag_index].value = has_key(PHOTILS.selected_tags, tag)
+
+        if tag then
+            GUI.tag_box[tag_index].label = tag
+            GUI.tag_box[tag_index].sensitive = true
+        else
+            GUI.tag_box[tag_index].label = ""
+            GUI.tag_box[tag_index].sensitive = false
+        end
+        tag_index = tag_index + 1
+    end
+
+    PHOTILS.in_pagination = false
+end
+
+function PHOTILS.attach_tags()
+    local image = dt.gui.selection()[1]
+    for tag, _ in pairs(PHOTILS.selected_tags) do
+        local dt_tag = dt.tags.create(tag)
+        dt.tags.attach(dt_tag, image)
+    end
+
+    dt.print(_("Tags successfully attached to image"))
+end
+
+function PHOTILS.get_tmp_file()
+    local tmp_file = os.tmpname()
+    if dt.configuration.running_os == "windows" then
+        tmp_file = dt.configuration.tmp_dir .. tmp_file -- windows os.tmpname() defaults to root directory
+    end
+
+    local f = io.open(tmp_file, "w")
+    if not f then
+        dt.print_log(string.format(_("Error writing to `%s`"), tmp_file))
+        os.remove(tmp_file)
+        return nil
+    end
+
+    return tmp_file
+end
+
+function PHOTILS.get_tags(image, with_export)
+    local tmp_file = PHOTILS.get_tmp_file()
+    local in_arg = df.sanitize_filename(tostring(image))
+    local out_arg = df.sanitize_filename(tmp_file)
+    local executable = photils_installed
+
+    if dt.configuration.running_os == "macos" then
+        executable =  executable .. "/Contents/MacOS/photils-cli"
```

Comment:
I have no clue about macOS. Is it save to use a fixed path?

Reply:
@supertobi I think this is safe since @scheckmedia controls the executable and where it's installed.